core/vm/gas_table.go

```diff
@@ -0,0 +1,294 @@
+}
+
+func gasCreate(gt params.GasTable, env *Environment, contract *Contract, stack *Stack, mem *Memory, memorySize *big.Int) *big.Int {
+	return new(big.Int).Add(gasTable[CREATE], memoryGasCost(mem, memorySize))
+}
+
+func gasCall(gt params.GasTable, env *Environment, contract *Contract, stack *Stack, mem *Memory, memorySize *big.Int) *big.Int {
+	gas := new(big.Int).Set(gt.Calls)
+
+	transfersValue := stack.Back(2).BitLen() > 0
+	var (
+		address = common.BigToAddress(stack.Back(1))
+		eip158  = env.ChainConfig().IsEIP158(env.BlockNumber)
+	)
+	if eip158 {
+		if env.StateDB.Empty(address) && transfersValue {
+			gas.Add(gas, params.CallNewAccountGas)
+		}
+	} else if !env.StateDB.Exist(address) {
+		gas.Add(gas, params.CallNewAccountGas)
+	}
+	if transfersValue {
+		gas.Add(gas, params.CallValueTransferGas)
+	}
+	gas.Add(gas, memoryGasCost(mem, memorySize))
+
+	cg := callGas(gt, contract.Gas, gas, stack.data[stack.len()-1])
+	// Replace the stack item with the new gas calculation. This means that
+	// either the original item is left on the stack or the item is replaced by:
+	// (availableGas - gas) * 63 / 64
+	// We replace the stack item so that it's available when the opCall instruction is
+	// called. This information is otherwise lost due to the dependency on *current*
+	// available gas.
+	stack.data[stack.len()-1] = cg
+
+	return gas.Add(gas, cg)
+}
+
+func gasCallCode(gt params.GasTable, env *Environment, contract *Contract, stack *Stack, mem *Memory, memorySize *big.Int) *big.Int {
+	gas := new(big.Int).Set(gt.Calls)
+	if stack.Back(2).BitLen() > 0 {
+		gas.Add(gas, params.CallValueTransferGas)
+	}
+	gas.Add(gas, memoryGasCost(mem, memorySize))
+
+	cg := callGas(gt, contract.Gas, gas, stack.data[stack.len()-1])
+	// Replace the stack item with the new gas calculation. This means that
+	// either the original item is left on the stack or the item is replaced by:
+	// (availableGas - gas) * 63 / 64
+	// We replace the stack item so that it's available when the opCall instruction is
+	// called. This information is otherwise lost due to the dependency on *current*
+	// available gas.
+	stack.data[stack.len()-1] = cg
+
+	return gas.Add(gas, cg)
+}
+
+func gasReturn(gt params.GasTable, env *Environment, contract *Contract, stack *Stack, mem *Memory, memorySize *big.Int) *big.Int {
+	return new(big.Int).Add(gasTable[RETURN], memoryGasCost(mem, memorySize))
+}
+
+func gasSuicide(gt params.GasTable, env *Environment, contract *Contract, stack *Stack, mem *Memory, memorySize *big.Int) *big.Int {
+	gas := new(big.Int)
+	// EIP150 homestead gas reprice fork:
+	if env.ChainConfig().IsEIP150(env.BlockNumber) {
+		gas.Set(gt.Suicide)
+		var (
+			address = common.BigToAddress(stack.Back(0))
+			eip158  = env.ChainConfig().IsEIP158(env.BlockNumber)
+		)
+
+		if eip158 {
+			// if empty and transfers value
+			if env.StateDB.Empty(address) && env.StateDB.GetBalance(contract.Address()).BitLen() > 0 {
```

Comment:
The uglyness of `env.StateDB.GetBalance(contract.Address()).BitLen() > 0` could be hidden within `env.StateDB.HasBalance(contract.Address)` , instead of having it all over the place. 

And if there is ever a 0-value-account-bloomfilter or whatever, you can swap it out in one place.